lib/commands/skill/add-locales/helper.js

```diff
@@ -0,0 +1,130 @@
+    }
+    const manifestPath = path.join(skillPackageSrc, CONSTANTS.FILE_PATH.SKILL_PACKAGE.MANIFEST);
+    new Manifest(manifestPath);
+}
+
+/**
+ * Add locales by deciding where to get the source of interaction models,
+ * and commit the copy paste from the source to the target location.
+ *
+ * @param {Array} selectedLocales
+ * @param {String} profile
+ * @param {Boolean} doDebug
+ * @param {Function} callback
+ */
+function addLocales(selectedLocales, profile, doDebug, callback) {
+    const iModelSourceByLocales = _resolveSourcesForIModel(selectedLocales, profile, doDebug);
+    const uniqueTemplateSet = new Set(iModelSourceByLocales.values());
+    const iModelFolderPath = path.join(process.cwd(), ResourcesConfig.getInstance().getSkillMetaSrc(profile),
+        CONSTANTS.FILE_PATH.SKILL_PACKAGE.INTERACTION_MODEL, 'custom');
+    // copy iModel JSONs that contains same language
+    iModelSourceByLocales.forEach((v, k) => {
+        if (fs.existsSync(v)) {
+            const sourceLocale = path.basename(v, path.extname(v));
+            const targetFilePath = path.join(iModelFolderPath, `${k}.json`);
+            fs.copySync(v, targetFilePath); // do not fail if the target locale exists already
+            uniqueTemplateSet.delete(v);
+            const sourceManifestLocale = Manifest.getInstance().getPublishingLocale(sourceLocale);
+            Manifest.getInstance().setPublishingLocale(k, sourceManifestLocale);
+        }
+    });
+    // retrieve remote templates when no existing JSON shares the same language
+    _retrieveTemplatesByLanguage(uniqueTemplateSet, doDebug, (templateErr, templateByLanguage) => {
+        if (templateErr) {
+            return callback(templateErr);
+        }
+        iModelSourceByLocales.forEach((v, k) => {
+            if (!fs.existsSync(v)) {
+                const targetFilePath = path.join(iModelFolderPath, `${k}.json`);
+                fs.writeFileSync(targetFilePath, templateByLanguage.get(v), { encoding: 'utf-8' });
+                Manifest.getInstance().setPublishingLocale(k, { name: 'please change' });
+            }
+        });
+        Manifest.getInstance().write();
+        callback(undefined, iModelSourceByLocales);
+    });
+}
+
+/**
+ * Based on user's selected locales, and the local project's iModel files, decide where to get the
+ * iModel file to add in user's local project. Exclude reusableLocales themselves from the result.
+ *
+ * @param {Array} selectedLocales
+ * @param {String} profile
+ * @param {Boolean} doDebug
+ * @returns {Map} { locale: filePath/remoteURI }
+ */
+function _resolveSourcesForIModel(selectedLocales, profile, doDebug) {
+    const skillMetaController = new SkillMetadataController({ profile, doDebug });
+    const localIModelByLocale = skillMetaController.getInteractionModelLocales();
+    const result = new Map();
+    for (const locale of selectedLocales) {
+        if (R.keys(localIModelByLocale).includes(locale)) {
+            continue;
```

Comment:
I think it would be be much cleaner to write more functional style using find and map

Reply:
You can also write without using rambda. For example

```
const skillMetaController = new SkillMetadataController({ profile, doDebug });
    const currentLocales = skillMetaController.getInteractionModelLocales();
    const currentLocalesKeys = Object.keys(currentLocales);
    const getLangCode = (s) => R.slice(0, R.lastIndexOf('-', s), s);
    const newLocales = selectedLocales
        .filter(locale => !currentLocales[locale])
        .map(locale => {
            const lang = getLangCode(locale);
            const reusableLocaleKey = currentLocalesKeys.find(l => getLangCode(l) === lang);
            if (reusableLocaleKey) {
                return [locale, currentLocales[reusableLocaleKey]];
            }
            return [locale, CONSTANTS.TEMPLATES.INTERACTION_MODEL_BY_LANGUAGE[lang]];
        });
    return new Map(newLocales);
```